csrc/scheduler/normalization.cpp

```diff
@@ -516,18 +516,18 @@ std::shared_ptr<ReductionParams> innerPersistentHeuristic(
   const int64_t threads_after_vectorize =
       inner_most_dimension_numel / inner_reduction_unroll_factor;
   constexpr int64_t scheduler_per_sm = 4;
+  const int64_t min_threads_per_block = scheduler_per_sm * dev_prop->warpSize;
   if (outer_reduction_numel == 1 && vectorize) {
-    bdimx = std::min(
-        scheduler_per_sm * dev_prop->warpSize, threads_after_vectorize);
+    bdimx = std::min(min_threads_per_block, threads_after_vectorize);
   }
 
-  // If we don't have a full warp, let's do multiple reductions per block.
-  // Still keep vectorization as it is important for performance since V100.
-  // Limit block size to 4 warps to avoid occupancy and SM wave tail issues.
-  if (bdimx * bdimy * bdimz < warp_size) {
+  // If we don't have enough threads, let's do multiple reductions per block.
+  // Multiple reductions per block shows better performance than unroll
+  // iterations. Still keep vectorization as it is important for performance
+  // since V100.
+  if (bdimx * bdimy * bdimz < min_threads_per_block) {
     bdimy = std::min(
-        scheduler_utils::safeDiv(
-            scheduler_per_sm * dev_prop->warpSize, bdimx * bdimz),
```

Comment:
This previous heuristic seems strange to me. I remember `warp_size` is not `32` but some small number. Since you showed a pretty slow case with a block size of `(1, 1, 1)`, was `warp_size` `1` in that case?

In any case, the if condition uses `warp_size` but the actual adjust ment uses `scheduler_per_sm * dev_prop->warpSize`, which seems inconsistent.

Reply:
that's right. I feel the calculated non-32 `warp_size` doesn't help on performance, at least for A100. Maybe we should remove it. And simply follow the classic recommendation of 128 as the minimum thread block size.